### ISSUE TITLE
gcc 9 -Wdeprecated-copy

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#4166]
[-Wdeprecated-copy](https://gcc.gnu.org/onlinedocs/gcc-9.1.0/gcc/C_002b_002b-Dialect-Options.html#index-Wdeprecated-copy) was added in gcc 9 and enabled as part of `-Wextra`: https://gcc.gnu.org/gcc-9/changes.html
```
New warnings:
    -Wdeprecated-copy, implied by -Wextra, warns about the C++11 deprecation of implicitly declared copy constructor and assignment operator if one of them is user-provided. -Wdeprecated-copy-dtor also warns if the destructor is user-provided, as specified in C++11.
```

This new warning is raised in some places ([example](https://pastebin.com/RHgG9iTB) reported by @ivanovic).

If nothing else this would cause problems with Wesnoth's travis builds since the strict jobs fail when any warnings are reported.

**Comment by CelticMinstrel** [comment]
Fixing these warnings just involves adding a defaulted declaration, right?

**Comment by Pentarctagon** [comment]
I don't really know, and I don't have gcc 9 installed to be able to check either.  Mostly I opened this just to document it as happening ahead of gcc 9 becoming more commonly used.

**Comment by stevecotton** [comment]
I do have gcc-9 installed, and am checking the warnings now. Adding a defaulted declaration seems sufficient.

IMO, wfl::variant's handling of const needs a bit more documentation. The existing operator=(const& v) gives a non-const instance pointing to the same value_ object as v, but AFAICS that's ok because variant provides only immutable access to value_'s value.

The movetype class also needs changes, for which I have a WIP in #2486. I'll get that cleaned up and open a PR for it.

**Comment by stevecotton** [comment]
Does anyone know why gui::widget's copy constructor calls events::sdl_handler's default constructor, instead of copy-constructing that class from `o`'s base-class? This line was last modified in b308df8c7eb, and AFAICS it means the copy may be listening for events that the original isn't listening for.

https://github.com/wesnoth/wesnoth/blob/5ef720e9dde17baff1755894ef86e801170df427/src/widgets/widget.cpp#L33-L39

Pinging @aginor

**Comment by jyrkive** [comment]
It looks like `events::sdl_handler` is not designed to be copied to begin with. I think the code you found is a mistake.

**Comment by stevecotton** [comment]
Commit 05d109215d has silenced the warnings. I think we still need to look in more depth at the code - the warning was added because the compiler suspects there are bugs when the default implementation is used.

**Comment by loonycyborg** [comment]
In that commit I was able to solve it mostly by removing copy-constructors. Only two harder cases, one addressed by https://github.com/wesnoth/wesnoth/pull/4169 , other one in movetype code. Maybe it needs closer look too, but it's less obvious than first case with topic_text

**Comment by stevecotton** [comment]
Fixed on both master and 1.14.
